drivers/sdhc/sdhc_stm32.c

```diff
@@ -0,0 +1,797 @@
+			sys_cache_data_invd_range(dev_data->sdio_dma_buf,
+						dev_data->total_transfer_bytes);
+			memcpy(data->data, dev_data->sdio_dma_buf, data->block_size * data->blocks);
+		}
+
+		k_free(dev_data->sdio_dma_buf);
+	}
+
+	return res == HAL_OK ? 0 : -EIO;
+}
+
+static int sdhc_stm32_switch_to_1_8v(const struct device *dev)
+{
+	uint32_t res;
+	struct sdhc_stm32_data *data = dev->data;
+	const struct sdhc_stm32_config *config = dev->config;
+
+	/* Check if host supports 1.8V signaling */
+	if (!data->props.host_caps.vol_180_support) {
+		LOG_ERR("Host does not support 1.8V signaling");
+		return -ENOTSUP;
+	}
+
+	res = SDMMC_CmdVoltageSwitch(config->hsd->Instance);
+	if (res != 0) {
+		LOG_ERR("CMD11 failed: %#x", res);
+		return -EIO;
+	}
+
+	LOG_DBG("Successfully switched to 1.8V signaling");
+	return 0;
+}
+
+static int sdhc_stm32_request(const struct device *dev, struct sdhc_command *cmd,
+			      struct sdhc_data *data)
+{
+	int res = 0;
+	uint32_t sdmmc_res = 0U;
+	struct sdhc_stm32_data *dev_data = dev->data;
+	const struct sdhc_stm32_config *config = dev->config;
+
+	__ASSERT_NO_MSG(cmd != NULL);
+
+	if (k_mutex_lock(&dev_data->bus_mutex, K_MSEC(cmd->timeout_ms))) {
+		return -EBUSY;
+	}
+
+	if (HAL_SDIO_GetState(config->hsd) != HAL_SDIO_STATE_READY) {
+		LOG_ERR("SDIO Card is busy");
+		k_mutex_unlock(&dev_data->bus_mutex);
+		return -ETIMEDOUT;
+	}
+
+	(void)pm_device_runtime_get(dev);
+
+	/* Prevent the clocks to be stopped during the request */
+	pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+
+	dev_data->cmd_index = cmd->opcode;
+	switch (cmd->opcode) {
+	case SD_GO_IDLE_STATE:
+		sdmmc_res = sdhc_stm32_go_idle_state(dev);
+		if (sdmmc_res != 0U) {
+			res = -EIO;
+		}
+		break;
+
+	case SD_SELECT_CARD:
+		sdmmc_res = SDMMC_CmdSelDesel(config->hsd->Instance, cmd->arg);
+		if (sdmmc_res != 0U) {
+			res = -EIO;
+			break;
+		}
+		/* Clear unused flags to avoid SDIO card identification issues */
+		cmd->response[0U] &=
+			~(SD_R1_ERASE_SKIP | SD_R1_CSD_OVERWRITE | SD_R1_ERASE_PARAM);
+		break;
+
+	case SD_SEND_RELATIVE_ADDR:
+		sdmmc_res = SDMMC_CmdSetRelAdd(config->hsd->Instance, (uint16_t *)&cmd->response);
+		if (sdmmc_res != 0U) {
+			res = -EIO;
+			break;
+		}
+		/*
+		 * Restore RCA by reversing the double 16-bit right shift from
+		 * Zephyr subsys and SDMMC_CmdSetRelAdd
+		 */
+		cmd->response[0] = cmd->response[0] << 16;
+		break;
+
+	case SDIO_SEND_OP_COND:
+		sdmmc_res = SDMMC_CmdSendOperationcondition(config->hsd->Instance, cmd->arg,
+								(uint32_t *)&cmd->response);
+		if (sdmmc_res != 0U) {
+			res = -EIO;
+		}
+		break;
+
+	case SDIO_RW_DIRECT:
+		res = sdhc_stm32_rw_direct(dev, cmd);
+		break;
+
+	case SDIO_RW_EXTENDED:
+		__ASSERT_NO_MSG(data != NULL);
+		res = sdhc_stm32_rw_extended(dev, cmd, data);
+		break;
+
+	case SD_VOL_SWITCH:
+		res = sdhc_stm32_switch_to_1_8v(dev);
+		break;
+
+	default:
+		LOG_DBG("Unsupported Command, opcode:%d", cmd->opcode);
```

Comment:
Non-blocking, but I might change this to a `LOG_ERR`. Users should have a relatively easy way to see that this driver only supports the SDIO stack.